File: exercicios/projeto-final/skeletons/rickandmorty/kmp/pratica/androidApp/src/main/kotlin/com/puciec/rickmortykmp/FavoritesStore.kt
package com.puciec.rickmortykmp

import android.content.Context

private const val PREFS_NAME = "favorites"
private const val KEY_IDS = "favorite_character_ids"

// Android-specific no skeleton: SharedPreferences é a persistência local da
// plataforma, e guardamos apenas IDs para buscar os dados atuais na API.
class FavoritesStore(context: Context) {
    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    fun load(): Set<Int> =
        prefs.getStringSet(KEY_IDS, emptySet())?.mapNotNull { it.toIntOrNull() }?.toSet() ?: emptySet()

    fun toggle(id: Int): Set<Int> {
        val current = load().toMutableSet()
        if (!current.add(id)) current.remove(id)
        prefs.edit().putStringSet(KEY_IDS, current.map { it.toString() }.toSet()).apply()
        return current
    }
}

File: exercicios/projeto-final/skeletons/themealdb/flutter/pratica/lib/storage/favorites_store.dart
import 'package:shared_preferences/shared_preferences.dart';

/// Persistência local simples dos ids de receitas favoritadas.
class FavoritesStore {
  static const _key = 'favorite_meal_ids';

  Future<Set<String>> load() async {
    final prefs = await SharedPreferences.getInstance();
    final stored = prefs.getStringList(_key) ?? [];
    return stored.toSet();
  }

  Future<void> toggle(String id) async {
    final prefs = await SharedPreferences.getInstance();
    final current = await load();
    if (current.contains(id)) {
      current.remove(id);
    } else {
      current.add(id);
    }
    await prefs.setStringList(_key, current.toList());
  }
}
